src/main/java/com/zendesk/maxwell/producer/MaxwellNatsProducer.java

```diff
@@ -0,0 +1,229 @@
+ */
+public class MaxwellNatsProducer  extends AbstractProducer {
+	private static final Logger logger = LoggerFactory.getLogger(MaxwellNatsProducer.class);
+
+	private final ArrayBlockingQueue<RowMap> queue;
+	private MaxwellNatsProducerWorker worker;
+
+	public MaxwellNatsProducer(MaxwellContext context, String natsSubject, String clusterId, String clientId) {
+		super(context);
+
+		this.queue = new ArrayBlockingQueue<>(1000);
+		try {
+			this.worker = new MaxwellNatsProducerWorker(context, natsSubject, clusterId, clientId, queue);
+		} catch (Exception e) {
+			this.context.terminate(new RuntimeException(e));
+		}
+		Thread thread = new Thread(this.worker, "maxwell-nats-worker");
+		thread.setDaemon(true);
+		thread.start();
+	}
+
+	@Override
+	public void push(RowMap r) throws Exception {
+		this.queue.put(r);
+	}
+}
+
+class MaxwellNatsProducerWorker extends AbstractAsyncProducer implements Runnable {
+	static final Logger LOGGER = LoggerFactory.getLogger(MaxwellNatsProducer.class);
+
+	private final String natsSubject;
+
+	private Connection natsConnection;
+	private NatsConnectionListener connectionListener;
+	private StreamingConnection streamingConnection;
+	private final ArrayBlockingQueue<RowMap> queue;
+	private Thread thread;
+
+
+	public MaxwellNatsProducerWorker(MaxwellContext context, String natsSubject, String clusterId,
+					 String clientId, ArrayBlockingQueue<RowMap> queue) throws IOException, InterruptedException {
+		super(context);
+
+		this.natsSubject = natsSubject;
+		this.queue = queue;
+		this.connectionListener = new NatsConnectionListener(context);
+
+		String natsUrl = context.getConfig().natsUrl;
+		Long pubAckTimeout = context.getConfig().natsPubAckTimeout;
+
+		Options.Builder clientOptions = new Options.Builder();
+		io.nats.streaming.Options.Builder streamingOptions = new io.nats.streaming.Options.Builder();
+
+		clientOptions
+			.server(natsUrl)
+			.connectionListener(connectionListener);
+
+		if (context.getConfig().natsMaxReconnects > 0) {
+			clientOptions.maxReconnects(context.getConfig().natsMaxReconnects);
+		}
+
+		if (context.getConfig().natsReconnectWait != null) {
+			clientOptions.reconnectWait(Duration.ofMillis(context.getConfig().natsReconnectWait));
+		}
+
+		this.natsConnection = Nats.connect(clientOptions.build());
+		streamingOptions.natsConn(natsConnection);
+
+		if (pubAckTimeout != null) {
+			streamingOptions.pubAckWait(Duration.ofMillis(pubAckTimeout));
+		}
+
+		this.streamingConnection = NatsStreaming.connect(clusterId, clientId, streamingOptions.build());
+
+	}
+
+	@Override
+	public void run() {
+		this.thread = Thread.currentThread();
+		while ( true ) {
+			try {
+				//Irrespective of Nats is working or not, heartbeat should be pushed
+				if (queue.peek() instanceof HeartbeatRowMap) {
+					RowMap row = queue.take();
+					this.push(row);
+				} else if (connectionListener.isNatsRunning()){
+					RowMap row = queue.take();
+					this.push(row);
+				}
```

Comment:
this will be a very tight loop if nats is in a not-yet-connected state, we probably want to sleep for a few milliseconds if we're waiting for the connection to come online, no?